torch/csrc/autograd/python_variable.cpp

```diff
@@ -200,7 +200,11 @@ static PyObject *THPVariable_is_leaf(THPVariable *self)
 static PyObject * THPVariable_get_data(THPVariable *self)
 {
   HANDLE_TH_ERRORS
-  return THPVariable_Wrap(make_variable(self->cdata.data(), false));
+  auto var = make_variable(self->cdata.data(), false);
+  /// NOTE: we need to set `allow_tensor_metadata_change_` to false, to prevent users from
+  /// changing size or storage of `tensor.data`, because those changes will not update `tensor`.
+  var.data().unsafeGetTensorImpl()->set_allow_tensor_metadata_change(false);
```

Comment:
can you remind me why we don't need to do this at the C++ level for `.data`?

Reply:
C++'s `.data()` function doesn't need this because `.data()` in C++ always returns the same tensor as the Variable's underlying tensor and doesn't do any shallow copying, i.e. updating metadata on the tensor obtained from C++ `.data()` will always update the original tensor.

Reply:
another way of saying this is C++ `.data()` returns a Tensor, while python `.data` returns a Variable, correct?

Reply:
Yes that's correct